### DISCUSSION TITLE
loadDecodedBuffer in version 7

### DISCUSSION THREAD

**Discussion by vishal-4way** [katspaugh/wavesurfer.js#3201] [Q&A] [answered]
hello all

I already checked the https://github.com/katspaugh/wavesurfer.js/discussions/2806

while using >
const url = URL.createObjectURL(new Blob([audioBuffer], { type: 'audio/wav' }))
const peaks = [ audioBuffer.getChannelData(0), audioBuffer.getChannelData(1) ]
const duration = audioBuffer.duration

wavesurfer.load(url, peaks, duration);

I am getting this error (Uncaught (in promise) DOMException: The element has no supported sources.) at 
wavesurfer7.js:1

Is there any alternative way for using loadDecodedBuffer as because of using blob don't no why Iphone increase the total resources size, single audio file becomes two for the browser using blog

and you guys are planning for adding loadDecodedBuffer method in the version 7 ?

thanks!

**Comment by katspaugh** [comment]
You could try `loadBlob`:

```js
const blob = new Blob([audioBuffer], { type: 'audio/wav' })
const peaks = [ audioBuffer.getChannelData(0), audioBuffer.getChannelData(1) ]
const duration = audioBuffer.duration
wavesurfer.loadBlob(blob, peaks, duration);
```

**Reply by vishal-4way** [reply]
Still  I am getting this error (Uncaught (in promise) DOMException: The element has no supported sources.) at
wavesurfer7.js:1

**Reply by katspaugh** [reply]
Show your code.
If you read #2806, there's an edit saying you need to also set the header bytes as per https://stackoverflow.com/questions/62172398/convert-audiobuffer-to-arraybuffer-blob-for-wav-download/62173861#62173861

**Reply by vishal-4way** [reply]
thanks ! 
after applying this  https://stackoverflow.com/questions/62172398/convert-audiobuffer-to-arraybuffer-blob-for-wav-download/62173861#62173861

we can play the songs thanks !

Is there any alternative way for using loadDecodedBuffer as because of using blob don't no why Iphone increase the total resources size, single audio file becomes two for the browser using blog

and you guys are planning for adding loadDecodedBuffer method in the version 7 ?

**Reply by katspaugh** [reply, marked as answer]
No plans to bring back `loadDecodedBuffer` as we're not using a Web Audio buffer for playback anymore.

**Reply by vishal-4way** [reply]
thanks!

**Reply by lin2019** [reply]
When loading, the waveform of a single channel can be displayed, but it quickly becomes a complete waveform again

**Reply by lin2019** [reply]
In other words, I have a dual channel audio, but I only want to display the left channel

**Reply by Mattk70** [reply]
@katspaugh wrote:
> You could try `loadBlob`:
> 
> ```js
> const blob = new Blob([audioBuffer], { type: 'audio/wav' })
> const peaks = [ audioBuffer.getChannelData(0), audioBuffer.getChannelData(1) ]
> const duration = audioBuffer.duration
> wavesurfer.loadBlob(blob, peaks, duration);
> ```

Since loadBlob calls [loadAudio](https://github.com/katspaugh/wavesurfer.js/blob/e519c20710ade4b116287c805db77e2e30d855c4/src/wavesurfer.ts#L422) internally and this function will create peaks and duration if not supplied, it seems for me there is a simpler solution. Based on your suggestion I had written:

```js
async function loadBuffer(audio){ //audio is an ArrayBuffer containing s16le audio with a wave header
        // Recreate TypedArray
        const int16Array = new Int16Array(audio.buffer);

        // Convert to Float32Array (Web Audio API uses Float32 samples)
        const float32Array = new Float32Array(int16Array.length);
        for (let i = 0; i < int16Array.length; i++) {
            float32Array[i] = int16Array[i] / 32768; // Normalize from Int16 to Float32
        }

        // Create AudioBuffer using AudioContext
        const audioContext = new AudioContext();
        const audioBuffer = audioContext.createBuffer(1, float32Array.length, sampleRate); // Mono channel
        // Populate the AudioBuffer with float32Array data
        audioBuffer.copyToChannel(float32Array, 0);
        const blob = new Blob([audio], { type: 'audio/wav' });
        const peaks = [ audioBuffer.getChannelData(0) ]
        const duration = audioBuffer.duration
        await wavesurfer.loadBlob(blob, peaks, duration);
}
```
However, the following works just fine:

```js
async function loadBuffer(audio){
        const blob = new Blob([audio], { type: 'audio/wav' });
        await wavesurfer.loadBlob(blob);
}
```

**Comment by Guillermogsjc** [comment]
hello @katspaugh,

and what happens if you want to provide an ArrayBuffer representing the wav as it was done in v6 like this
for a componente of type WebAudio

const audioContext = new (window.AudioContext)();
const audioBuffer = await audioContext.decodeAudioData(elementArrayBuffer);
this.singleWave.loadDecodedBuffer(audioBuffer);
in this example there is no blob or explicit uri generated. Maybe you want to avoid the browser showing the audio blob or any uri with the audio

In the example use case i have an audio encrypted, and I decrypt it into an ArrayBuffer and then provide it to wavesurfer component as stated, being opaque for the console and network the decrypted audio

By dropping this loadDecodedBuffer and force the blob way, you lose this use case, not clear why.

On the use case exposed to access the decrypted audio you need to go against the memory directly, it is several points harder than exposing it thorugh a blob or a URL.

**Reply by katspaugh** [reply]
v7 is no longer Web Audio-centric, and that's the reason to not support this use case. Wavesurfer wants a URL, a blob, or a media element as its input.

The Web Audio implementation is v7 is an emulation of the HTMLMediaElement interface: https://raw.githubusercontent.com/katspaugh/wavesurfer.js/main/src/webaudio.ts. You can copy it to make your own pre-decoded buffer-based custom media element.

**Reply by Guillermogsjc** [reply]
thanks @katspaugh for the clue to implement the pre-decoded buffer

**Reply by Lonli-Lokli** [reply]
I need to load reversed audio without losing a quality, unfortunately with Blob it happens
